File: Autolock/Storage/LogStore.swift
import Foundation

struct LogEntry: Codable, Identifiable {
    let id: UUID
    let timestamp: Date
    let event: String
    let detail: String

    init(event: String, detail: String = "", timestamp: Date = Date()) {
        self.id = UUID()
        self.timestamp = timestamp
        self.event = event
        self.detail = detail
    }
}

/// Append-only structured log, persisted to a JSON file in the app's
/// Application Support directory so it survives app relaunches and is
/// readable from the Settings screen without Xcode attached.
///
/// Deliberately never logs raw credential values -- callers pass in
/// human-readable summaries (e.g. "auth succeeded", "lock command sent")
/// rather than request/response bodies.
final class LogStore: @unchecked Sendable {
    static let shared = LogStore()

    private let maxEntries = 500
    private let fileURL: URL
    private let queue = DispatchQueue(label: "com.sendprobe.autolock.logstore")

    init() {
        let supportDir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        try? FileManager.default.createDirectory(at: supportDir, withIntermediateDirectories: true)
        self.fileURL = supportDir.appendingPathComponent("autolock-log.json")
    }

    func append(event: String, detail: String = "") {
        queue.sync {
            var entries = readAllLocked()
            entries.append(LogEntry(event: event, detail: detail))
            if entries.count > maxEntries {
                entries.removeFirst(entries.count - maxEntries)
            }
            writeLocked(entries)
        }
    }

    func all() -> [LogEntry] {
        queue.sync { readAllLocked().sorted { $0.timestamp > $1.timestamp } }
    }

    func clear() {
        queue.sync { writeLocked([]) }
    }

    private func readAllLocked() -> [LogEntry] {
        guard let data = try? Data(contentsOf: fileURL) else { return [] }
        return (try? JSONDecoder().decode([LogEntry].self, from: data)) ?? []
    }

    private func writeLocked(_ entries: [LogEntry]) {
        guard let data = try? JSONEncoder().encode(entries) else { return }
        try? data.write(to: fileURL, options: .atomic)
    }
}

File: android/app/src/main/java/com/sendprobe/autolock/storage/LogStore.kt
package com.sendprobe.autolock.storage

import android.content.Context
import com.sendprobe.autolock.model.LogEntry
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File

/**
 * Append-only structured log, persisted to a JSON file in app-private
 * storage so it survives process death and is readable from the in-app
 * log screen without a debugger attached.
 *
 * Deliberately never logs raw credential values -- callers pass in
 * human-readable summaries (e.g. "auth succeeded", "lock command sent")
 * rather than request/response bodies.
 */
class LogStore private constructor(private val file: File) {

    private val json = Json { ignoreUnknownKeys = true }
    private val lock = Any()

    fun append(event: String, detail: String = "") {
        synchronized(lock) {
            val entries = readAllLocked().toMutableList()
            entries.add(LogEntry(event = event, detail = detail))
            val trimmed = if (entries.size > MAX_ENTRIES) entries.takeLast(MAX_ENTRIES) else entries
            writeLocked(trimmed)
        }
    }

    fun all(): List<LogEntry> =
        synchronized(lock) { readAllLocked().sortedByDescending { it.timestamp } }

    fun clear() {
        synchronized(lock) { writeLocked(emptyList()) }
    }

    private fun readAllLocked(): List<LogEntry> {
        if (!file.exists()) return emptyList()
        return runCatching { json.decodeFromString<List<LogEntry>>(file.readText()) }.getOrDefault(emptyList())
    }

    private fun writeLocked(entries: List<LogEntry>) {
        runCatching { file.writeText(json.encodeToString(entries)) }
    }

    companion object {
        private const val MAX_ENTRIES = 500
        private const val FILE_NAME = "autolock-log.json"

        @Volatile private var instance: LogStore? = null

        fun get(context: Context): LogStore =
            instance ?: synchronized(this) {
                instance ?: LogStore(File(context.applicationContext.filesDir, FILE_NAME)).also { instance = it }
            }
    }
}
